testing/upgradetest/versions.go

```diff
@@ -36,33 +41,102 @@ var (
+
+func setAgentVersions() {
+	v, err := getAgentVersions()
 	if err != nil {
-		return nil, fmt.Errorf("error retrieving versions from Artifact API: %w", err)
+		panic(err)
 	}
-	if len(vList.Versions) == 0 {
-		return nil, errors.New("retrieved versions list from Artifact API is empty")
+	agentVersions = v
+}
+
+func getAgentVersions() (*AgentVersions, error) {
+	var filePath string
+	dir, err := os.Getwd()
+	if err != nil {
+		return nil, fmt.Errorf("failed to get the current directory: %w", err)
+	}
+	for {
+		dir = filepath.Dir(dir)
+		pathToCheck := filepath.Join(dir, AgentVersionsFilename)
+		fi, err := os.Stat(pathToCheck)
+		if (err == nil || os.IsExist(err)) && !fi.IsDir() {
+			filePath = pathToCheck
+			break
+		}
+		if strings.HasSuffix(dir, string(filepath.Separator)) {
+			return nil, fmt.Errorf("failed to find %s", AgentVersionsFilename)
+		}
 	}
-	err = aac.RemoveUnreleasedVersions(ctx, vList)
+
+	f, err := os.OpenFile(filePath, os.O_RDONLY, 0)
 	if err != nil {
-		return nil, fmt.Errorf("failed to remove unreleased versions: %w", err)
+		return nil, fmt.Errorf("failed to open %s", filePath)
 	}
+	defer f.Close()
 
-	return getUpgradableVersions(ctx, vList, upgradeToVersion, currentMajorVersions, previousMajorVersions)
+	d := json.NewDecoder(f)
+	var versionFile AgentVersions
+	err = d.Decode(&versionFile)
+	if err != nil {
+		return nil, fmt.Errorf("failed to decode JSON in %s", filePath)
+	}
+
+	return &versionFile, nil
 }
 
-// Internal version of GetUpgradableVersions() with the artifacts API dependency removed for testing.
-func getUpgradableVersions(ctx context.Context, vList *tools.VersionList, upgradeToVersion string, currentMajorVersions int, previousMajorVersions int) ([]*version.ParsedSemVer, error) {
-	parsedUpgradeToVersion, err := version.ParseVersion(upgradeToVersion)
-	if err != nil {
-		return nil, fmt.Errorf("upgradeToVersion %q is not a valid version string: %w", upgradeToVersion, err)
+// FetchUpgradableVersions returns the versions list from the agent version file.
+func GetUpgradableVersions() ([]*version.ParsedSemVer, error) {
+	ensureAgentVersions()
+
+	parsedVersions := make([]*version.ParsedSemVer, 0, len(agentVersions.TestVersions))
+	for _, v := range agentVersions.TestVersions {
+		parsed, err := version.ParseVersion(v)
+		if err != nil {
+			return nil, fmt.Errorf("failed to parse version %q from %s", v, AgentVersionsFilename)
+		}
+		parsedVersions = append(parsedVersions, parsed)
 	}
 
-	currentMajor := parsedUpgradeToVersion.Major()
-	var currentMajorSelected, previousMajorSelected int
+	return parsedVersions, nil
+}
+
+// FetchUpgradableVersions returns a list of versions that meet the specified requirements.
+//
+// Every version on the resulting list will meet the given requirements (by OR condition).
+// However, it's not guaranteed that the list contains the amount of versions per requirement.
+// For example, if only 2 previous minor versions exist but 5 requested, the list will have only 2.
```

Comment:
Would it be better to return an error in this case? Requesting a specific number of versions is not a requirement if it is actually considered optional by the implementation.

Reply:
@cmacknz this is going to be a reply to your comments above too: https://github.com/elastic/elastic-agent/pull/4331#issuecomment-1977200360 and https://github.com/elastic/elastic-agent/pull/4331#issuecomment-1977224749

Every time there is a different output from the mage target we're going to have a PR for a review (unless there is an existing one). So, there will be a CI build where we can see if the versions are used properly, it's up to us when to approve and merge those PRs manually.

The purpose of the generator is to create a file closest to our requirements, that we can review and approve.

We currently have all these issues too, we just don't see them, we don't even know what versions we're testing against in each run unless we check the detailed logs. This is the main problem this PR is solving.

Reply:
We can also fine-tune the version requirements in a follow up later as we see the incoming automation PRs.

Reply:
> I have a few questions about the output:
>
> Why do we list 8.13.0 instead of 8.13.0-SNAPSHOT? Because the build candidate exists?
> Why do we list 8.12.3? The latest minor release is 8.12.2 and there is no planned 8.12.3 yet.

In both cases these versions are [returned](https://artifacts-api.elastic.co/v1/versions/) by the artifact API as released and there is no additional check whether the artifact exists on the CDN or not.

<details>
<summary>response</summary>
```json
{
  "versions": [
    "7.17.14",
    "7.17.15",
    "7.17.16",
    "7.17.17",
    "7.17.18-SNAPSHOT",
    "7.17.18",
    "7.17.19-SNAPSHOT",
    "8.10.0",
    "8.10.1",
    "8.10.2",
    "8.10.3",
    "8.10.4",
    "8.11.0",
    "8.11.1",
    "8.11.2",
    "8.11.3",
    "8.11.4",
    "8.12.0",
    "8.12.1-SNAPSHOT",
    "8.12.1",
    "8.12.2-SNAPSHOT",
    "8.12.2",
    "8.12.3-SNAPSHOT",
    "8.12.3",
    "8.13.0-SNAPSHOT",
    "8.13.0",
    "8.14.0-SNAPSHOT"
  ],
  "aliases": [
    "7.17-SNAPSHOT",
    "7.17",
    "8.10",
    "8.11",
    "8.12-SNAPSHOT",
    "8.12",
    "8.13-SNAPSHOT",
    "8.13",
    "8.14-SNAPSHOT"
  ],
  "manifests": {
    "last-update-time": "Tue, 05 Mar 2024 13:33:02 UTC",
    "seconds-since-last-update": 161
  }
}
```
</details>

Reply:
OK so an 8.12.3 artifact does exist. Looking into https://artifacts-staging.elastic.co/elastic-agent-package/8.12.3-511be3e7/manifest-8.12.3.json and tacking down the job that produced it TIL that Buildkite now does a daily staging build for example https://buildkite.com/elastic/unified-release-staging/builds/813.

So a staging build for the next release patch release will always exist, even if we are not planning to actually release it. This is just a slower version of the snapshot build so we would never want to test both this and a snapshot, we don't gain enough additional test coverage. 

Every non-snapshot build return by the artifacts API is potentially an unreleased build candidate. I think we need a way to obviously tell official releases from build candidates when doing version selection. There are a few ways to do this:
1. We could use https://www.elastic.co/api/product_versions which returns every available release
2. We could try to download each version from artifacts.elastic.co to see if we get a 404.
3. We could look at the available releases on the elastic-agent GitHub page as Paolo suggested earlier.

I think for upgrades we want to always test the two previous minor versions, using the official releases when they are available so we simulate what real users would be doing. So for main, 8.13, and 8.12 currently we'd want to test:

* main: 8.14.0-SNAPSHOT, 8.13.0-SNAPSHOT, 8.12.2, 7.17.18
* 8.13: 8.13.0-SNAPSHOT, 8.12.2 8.11.4, and 7.17.18
* 8.12: 8.12.3-SNAPSHOT, 8.11.4, 8.10.4, 7.17.18

I am admittedly a bit torn on whether we should just always test the latest snapshot (e.g 8.12.3-SNAPSHOT) for a given branch or use the most recent public release if it exists (e.g. 8.12.2). I am currently biased towards using the public release since that is what will happen in reality when a branch is released.

Reply:
but wasn't the whole point of testing the latest snapshot to ensure upgrade isn't broken on the new version? Or we're covering that with that test which modifies the agent's package to simulate it's another version?

Reply:
There are two halves of the upgrade process to test:

1. We need to make sure that previous releases can upgrade to the version we are about to release. Previous releases need to be able to find the agent executable so they can replace themselves with it and ensure the new watcher works properly. For this we want to test the two previous minors and 7.17.
2. We need to make sure the version we are about to release can upgrade itself to another version. In practice it is a downgrade to a snapshot from the same version. This  tests logic to download, extract, and replace the binary in the current version.